rust/src/wrapper/device/mod.rs

```diff
@@ -230,35 +263,159 @@ impl Device {
     /// Returns the host used by the device, if any
     pub fn host(&mut self) -> PyResult<Option<Host>> {
         Python::with_gil(|py| {
-            self.0
+            self.obj
                 .getattr(py, intern!(py, "host"))
                 .map(|obj| obj.into_option(Host::from))
         })
     }
 
     /// Start advertising the data set with [Device.set_advertisement].
     pub async fn start_advertising(&mut self, auto_restart: bool) -> PyResult<()> {
+        if self.advertising_status == AdvertisingStatus::AdvertisingExtended {
+            return Err(PyErr::new::<PyException, _>("Already advertising in extended mode. Stop the existing extended advertisement to start a legacy advertisement."));
+        }
+
         Python::with_gil(|py| {
             let kwargs = PyDict::new(py);
             kwargs.set_item("auto_restart", auto_restart)?;
 
-            self.0
+            self.obj
                 .call_method(py, intern!(py, "start_advertising"), (), Some(kwargs))
                 .and_then(|coroutine| into_future(coroutine.as_ref(py)))
         })?
         .await
-        .map(|_| ())
+        .map(|_| ())?;
+
+        self.advertising_status = AdvertisingStatus::AdvertisingLegacy;
+        Ok(())
+    }
+
+    /// Start advertising the data set in extended mode, replacing any existing extended adv. The
+    /// advertisement will be non-connectable.
+    ///
+    /// Fails if the device is already advertising in legacy mode.
+    #[cfg(feature = "unstable_extended_adv")]
+    pub async fn start_advertising_extended(
+        &mut self,
+        adv_data: AdvertisementDataBuilder,
+    ) -> PyResult<()> {
+        // TODO: add tests when local controller object supports extended advertisement commands (github.com/google/bumble/pull/238)
+        match self.advertising_status {
+            AdvertisingStatus::AdvertisingLegacy => return Err(PyErr::new::<PyException, _>("Already advertising in legacy mode. Stop the existing legacy advertisement to start an extended advertisement.")),
+            // Stop the current extended advertisement before advertising with new data.
+            // We could just issue an LeSetExtendedAdvertisingData command, but this approach
+            // allows better future flexibility if `start_advertising_extended` were to change.
+            AdvertisingStatus::AdvertisingExtended => self.stop_advertising_extended().await?,
+            _ => {}
+        }
+
+        // if you change this, don't forget to change the same handle in `stop_advertising_extended`
+        let advertising_handle = 0x00;
```

Comment:
This will work temporarily for testing, but for real use cases, we'll need to support multiple advertising sets. For a typical device, you want to have one advertising set for legacy payloads (so that older peer devices can still find you) and one or more extended advertisers. Which does complicate the API somewhat (have to manage advertising instances, deal with running out of advertisers, etc.), which is why there's no official Bumble API for this yet (but working on it). The Android API is a good example for what an API could look like (even though it has its own gaps)

Reply:
Agreed that we will want to support multiple advertising sets in the future.

Once you've figured out how to support extended advertising in the core API, my plan is to completely drop this and adopt what you've implemented (which is why I've marked these additions as "unstable").

But in the meantime, we can at least use this for testing purposes.